packages/react/src/components/Masthead/MastheadTopNav.js

```diff
@@ -23,7 +27,68 @@ const { prefix } = settings;
  * Masthead top nav component.
  */
 const MastheadTopNav = ({ navigation, ...topNavProps }) => {
+  let containerWidth = 0;
+  const resizeObserver = new ResizeObserver(() => {
+    containerWidth =
+      calculateTotalWidth(['bx--header__search']) -
+      calculateTotalWidth(['bx--masthead__search']) -
+      calculateTotalWidth(['bx--header__name']);
+    const totalNavWidth = calculateTotalWidth(['bx--header__nav']);
+    if (totalNavWidth > containerWidth) {
+      setOverflow(true);
+      const offset = document.querySelector('.bx--header__nav-container')
+        .scrollLeft;
+      if (offset === 0 || offset + containerWidth <= totalNavWidth) {
+        setShowRightCaret(true);
+      }
+    } else {
+      setOverflow(false);
+      setShowLeftCaret(false);
+      setShowRightCaret(false);
+    }
+  });
+
+  useEffect(() => {
+    resizeObserver.observe(topNavProps.topNavRef.current);
```

Comment:
Any reason the DOM ref has to come from `<Masthead>` instead of `<div class="bx--header__nav-container">` that this component renders?

Reply:
The DOM ref has to come from `<Masthead>` because if the user starts in wider screen and then shrinks the size where the overflow should turn on then only the `<Masthead>` will change in size. An overflow style causes `<div class="bx--header__nav-container">` to stay the same size even if the user shrinks the browser, which will then not trigger the resizeObserver.